tsdb/head.go

```diff
@@ -774,7 +774,11 @@ func (h *Head) loadMmappedChunks(refSeries map[chunks.HeadSeriesRef]*memSeries)
 	if err := h.chunkDiskMapper.IterateAllChunks(func(seriesRef chunks.HeadSeriesRef, chunkRef chunks.ChunkDiskMapperRef, mint, maxt int64, numSamples uint16, encoding chunkenc.Encoding) error {
 		secondLastRef = lastRef
 		lastRef = chunkRef
-		isOOO := chunkenc.IsOutOfOrderChunk(encoding)
+		isOOO := h.chunkDiskMapper.IsOutOfOrderChunk(encoding)
+		// flip the first bit to restore the original encoding
+		if isOOO {
+			encoding = h.chunkDiskMapper.OutOfOrderDecoding(encoding)
```

Comment:
Hm I wonder what could be done here, it seems here the head is aware of OOO chunk encodings and this is precisely what we wanted to hide in the chunk disk mapper.

Maybe we could change the function to return a boolean saying if its OOO instead of the encoding and then if its OOO we add it to `oooMmappedChunks`  and if its not we add it to `mmappedChunks`

What do you think @codesome

Reply:
@jesusvazquez Can you point out as to which function you want to return `bool`.
The current function is part of the load chunks from initial read. I have tried to put all the logic in CDM. Do you want me to pass this `bool` as part of the callback function inside the `IterateAllChunks` ?

Reply:
Right! I believe I added the `encoding` to that function when we added the OOO chunks. If you notice the usages I think it should be fairly easy to change.

https://github.com/prometheus/prometheus/blob/3330d85ba8cd6c53e74af23ab96e9e2c134bec39/tsdb/chunks/head_chunks.go#L765

I would change that signature to something like
`func (cdm *ChunkDiskMapper) IterateAllChunks(f func(seriesRef HeadSeriesRef, chunkRef ChunkDiskMapperRef, mint, maxt int64, numSamples uint16, isOOO bool) error) (err error) {`

That will force you to refactor the call in https://github.com/prometheus/prometheus/blob/1f0cc09579b7d04bbcb2b9746b7040da55791d48/tsdb/head.go#L737 and remove the information about encodings in the head.

Reply:
+1 with what @jesusvazquez said. Let's change the function signature to return the isOOO boolean such that head doesn't need to infer it from the chunk encoding. The encoding coming out of ChunkDiskMapper here should have that ooo bit removed.

Reply:
Ok, will do